nn/transformer.py

```diff
@@ -0,0 +1,253 @@
+    :param norm_eps: Epsilon value for layer normalization
+    :param norm_first: if ``True`` will perform normalization before other att and ff operations, otherwise after
+    :param norm: normalization function
+    """
+    super().__init__()
+    self.self_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)
+    self.multihead_attn = MultiheadAttention(dim_model, num_heads, dropout=dropout)
+
+    self.linear_ff = nn.Linear(dim_ff)
+    self.linear_out = nn.Linear(dim_model)
+
+    self.norm = norm
+    self.norm_first = norm_first
+    self.norm_eps = norm_eps
+    self.activation = activation
+    self.dropout = dropout
+
+  def forward(self, tgt: nn.LayerRef, memory: nn.LayerRef) -> nn.LayerRef:
+    """
+    Two possible forward variants of decoder, defined by self.norm_first, tgt and memory have shape (B, T, F)
+    """
+    x = tgt
+    if self.norm_first:
+      x = x + self._self_attention_block(self.norm(x, epsilon=self.norm_eps))
+      x = x + self._multi_head_attention_block(self.norm(x, epsilon=self.norm_eps), memory)
+      x = x + self._feed_forward_block(self.norm(x, epsilon=self.norm_eps))
+    else:
+      x = self.norm(x + self._self_attention_block(x), epsilon=self.norm_eps)
+      x = self.norm(x + self._multi_head_attention_block(x, memory), epsilon=self.norm_eps)
+      x = self.norm(x + self._feed_forward_block(x), epsilon=self.norm_eps)
+
+    return x
+
+  def _self_attention_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.self_attn(x, x, x)
+    return nn.dropout(x, self.dropout)
+
+  def _multi_head_attention_block(self, x: nn.LayerRef, mem: nn.LayerRef) -> nn.LayerRef:
+    x = self.multihead_attn(x, mem, mem)
+    return nn.dropout(x, self.dropout)
+
+  def _feed_forward_block(self, x: nn.LayerRef) -> nn.LayerRef:
+    x = self.linear_ff(x)
+    x = self.activation(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    x = self.linear_out(x)
+    x = nn.dropout(x, dropout=self.dropout)
+    return x
+
+
+class TransformerDecoder(nn.Module):
+  """
+  Defines the full Decoder of the standard transformer
+  """
+  def __init__(self, decoder_layer: Union[TransformerDecoderLayer, Any], num_layers: int,
+               norm: Callable[[nn.LayerRef], nn.LayerRef] = nn.layer_norm, norm_eps: float = 1e-5):
+      """
+      :param decoder_layer: Decoder layer to be stacked num_layers times
+      :param num_layers: Number of layers
+      :param norm: normalization function for output layer normalization
+      :param norm_eps: Epsilon value for output layer normalization
+      """
+      super(TransformerDecoder, self).__init__()
+      import copy
+      self.layers = nn.ModuleList([copy.deepcopy(decoder_layer) for _ in range(num_layers)])
+
+      self.num_layers = num_layers
+      self.norm = norm
+      self.norm_eps = norm_eps
+
+  def forward(self, tgt: nn.LayerRef, memory: nn.LayerRef) -> nn.LayerRef:
+    """
+    Executes the decoder layer as often as in num_layers defined
```

Comment:
This comment is very misleading, or actually wrong.
- "Execute" is the wrong wrong term. It applies it.
- It does not apply the decoder layer multiple times, no. It applies all the decoder layer**s** once. Every decoder layer is a copy of the original decoder layer.